File: 119_PascalsTriangleII.py
# https://leetcode.com/problems/pascals-triangle-ii/

class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        # # v1 modified Q118
        # pascal=[]
        # for i in range(1,rowIndex+2):
        #     thisline=[1]*i
        #     if i > 2:
        #         for j in range(1,i-1):
        #             thisline[j]=pascal[i-2][j-1]+pascal[i-2][j]
        #     pascal.append(thisline)
        # return pascal[rowIndex]

        # v2 by the difinition of Pascal Triangle
        # Say we have the current layer [1, 2, 1]. 
        # We then make 2 copies of this layer, 
        # add 0 to the start of one copy, and add 0 to the end of one copy; 
        # then we have [0, 1, 2, 1] and [1, 2, 1, 0]. 
        # Then we can perform the element-wise add operation and we would have [1, 3, 3, 1]. This is from the definition of Pascal's Triangle.
        row = [1]
        for _ in range(rowIndex):
            row = [x + y for x, y in zip([0]+row, row+[0])]
        return row

File: LeetCode/119_PascalsTriangleII.java
// https://leetcode.cn/problems/pascals-triangle-ii
// 2022年11月09日 17:07:11
class Solution {
    public List<Integer> getRow(int rowIndex) {
        List<Integer> lastRow = new ArrayList<Integer>(Arrays.asList(1));
        if (rowIndex == 0) {
            return lastRow;
        }
        List<Integer> currRow = new ArrayList<Integer>(Arrays.asList(1, 1));
        for (int i = 2; i <= rowIndex; i++) {
            lastRow = currRow;
            currRow = new ArrayList<Integer>();
            currRow.add(1);
            for (int j = 1; j < i; j++) {
                currRow.add(lastRow.get(j - 1) + lastRow.get(j));
            }
            currRow.add(1);
        }
        return currRow;
    }
}